### PR TITLE
Ignore missing static libs if -H:+ExitAfterRelocatableFileWrite is set.

### LINKED ISSUE #1727: Allow developers to override location of static jdk libs
When there are no static JDK libs available, or when the developer prefers to use his own static libs (e.g. with different versions, or for different platforms), there is currently no option to override the default.

### PR DESCRIPTION
**Pull request by johanvos**
This fixes #1727

### CODE REVIEW

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageOptions.java [hunk 1]
```diff
@@ -198,4 +198,7 @@ public static int getMaximumNumberOfAnalysisThreads(OptionValues optionValues) {
         }
         return analysisThreads;
     }
+
+    @Option(help = "Specify a directory that contains static JDK libs (e.g. libjava.a etc)")//
+    public static final HostedOptionKey<String> StaticJDKLibs = new HostedOptionKey<>("");
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageOptions.java]: You can make this null by default.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java [hunk 2]
```diff
@@ -165,6 +166,13 @@ public TargetDescription getTarget() {
         Path staticLibsDir = null;
         String hint = null;
 
+        /* In case a specific location for the static JDK libraries is provided, use that. */
+
+        String externalStaticLibsDir = NativeImageOptions.StaticJDKLibs.getValue();
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java]: And use `hasBeenSet` here.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java [hunk 3]
```diff
@@ -165,6 +166,13 @@ public TargetDescription getTarget() {
         Path staticLibsDir = null;
         String hint = null;
 
+        /* In case a specific location for the static JDK libraries is provided, use that. */
+
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java]: Extra space.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java [hunk 4]
```diff
@@ -165,6 +166,13 @@ public TargetDescription getTarget() {
...
+        String externalStaticLibsDir = NativeImageOptions.StaticJDKLibs.getValue();
+        if (!externalStaticLibsDir.isEmpty()) {
+            staticLibsDir = Paths.get(externalStaticLibsDir);
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java]: And finally, use `getValue` here.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageOptions.java [hunk 5]
```diff
@@ -198,4 +198,7 @@ public static int getMaximumNumberOfAnalysisThreads(OptionValues optionValues) {
         }
         return analysisThreads;
     }
+
+    @Option(help = "Specify a directory that contains static JDK libs (e.g. libjava.a etc)")//
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageOptions.java]: libs -> libraries

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageOptions.java (hunk 1 above)
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageOptions.java]: I would also use `Libraries` as we already have it with `CLibraries`.

**Comment by johanvos** [comment]
> Instead of adding yet another hosted option I suggest we add checking if the specified com.oracle.svm.core.SubstrateOptions#CLibraryPath already contains the static libs. That means just run the same probing for all elements of `OptionUtils.flatten(",", SubstrateOptions.CLibraryPath.getValue())` in
> 
> https://github.com/oracle/graal/blob/519cb180a31c61f7735d64aeabf566236489ddcf/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/c/NativeLibraries.java#L184

There are actually 2 aspects to this:

1. In case -H:+ExitAfterRelocatableImageWrite is specified, you don't care about the static SDK libs at all. In that case, they won't be in CLibraryPath either. But this can be better dealt with by checking the ExitAfterRelocatableImageWrite first, and returning if that is set.

2. In case you are cross-compiling, you don't want to use the static libraries that you use at compiletime. In case the CLibraryPath is not used for that purpose (e.g. it does not contain native code relevant at compiletime), we can indeed reuse that option.

**Comment by olpaw** [comment]
> But this can be better dealt with by checking the ExitAfterRelocatableImageWrite first, and returning if that is set.

Yes that makes sense.

>     2. In case you are cross-compiling, you don't want to use the static libraries that you use at compiletime. In case the CLibraryPath is not used for that purpose (e.g. it does not contain native code relevant at compiletime), we can indeed reuse that option.

That is the case. None of the libs that are on `CLibraryPath` are _executed_ at image-build time. They are only used for _linking_ at image-build time. I.e. In the case of cross-compiling to - let's say - aarch64 on an amd64 host you would have aarch64 libraries on `CLibraryPath`.

**Comment by johanvos** [comment]
I modified the PR to satisfy goal #1 (don't bail if there are no static JDK libs found, but ExitAfterRelocatableImageWrite is set).
I propose to tackle goal #2 (allow to override the location) in a different PR, at a later moment.

**Pull request merged**